### ISSUE TITLE
Issue with Upload

### ISSUE THREAD

**Issue by aallard** [ClaesClaes/Arduino-ESP32-NimBLE-OTA-iOS-SwiftUI#3]
Here's the equipment I'm using:
* iPhone 15 running iOS 17.3.1
* ESP32 Dev Module WROOM 32E

I am using the NimBLE-Arduino library (v. 1.4.1) from h2zero and have made the config changes to "nimconfig.h" as you specified in the .ino file, with the exception that in version 1.4.1 the nimconfig.h file does not have the line:
"#define CONFIG_BT_NIMBLE_TASK_STACK_SIZE 8192". so that was not changed.  Otherwise the other defines were enabled.

And finally, my Arduino IDE settings look like this:
<img width="528" alt="Screenshot 2024-02-10 at 1 17 26 PM" src="https://github.com/ClaesClaes/Arduino-ESP32-NimBLE-OTA-iOS-SwiftUI/assets/8251857/66c2e2ce-476d-42bf-9d6f-4f1516fe539a">

All of the code compiles and runs on both platforms, and the iOS app makes a good connection to the ESP32.  When I click "send .bin to ESP32 over OTA" it does send data to the ESP32.  But, it will fail, with one of two errors.  I've attached the log output from the ESP32 to show what it says:

*** App connected
57:e7:93:de:6a:17
main loop started
1. BeginOTA
2. Running partition type 0 subtype 16 (offset 0x00010000) 
3. Writing to partition subtype 17 at offset 0x1f0000 
--Data received---
4. Final byte arrived
5. E (28465) esp_image: invalid segment length 0xffffffff
OTA end failed 
1. BeginOTA
2. Running partition type 0 subtype 16 (offset 0x00010000) 
3. Writing to partition subtype 17 at offset 0x1f0000 
E (34248) esp_ota_ops: OTA image has invalid magic byte (expected 0xE9, saw 0x35)
Error: write to flash failed
1. BeginOTA
2. Running partition type 0 subtype 16 (offset 0x00010000) 
3. Writing to partition subtype 17 at offset 0x1f0000 
E (40102) esp_ota_ops: OTA image has invalid magic byte (expected 0xE9, saw 0x42)
Error: write to flash failed
1. BeginOTA
2. Running partition type 0 subtype 16 (offset 0x00010000) 
3. Writing to partition subtype 17 at offset 0x1f0000 
E (45836) esp_ota_ops: OTA image has invalid magic byte (expected 0xE9, saw 0x65)
Error: write to flash failed
*** App disconnected
start advertising

I have tried running things with the sample.bin file that you provided, as well as creating a new .bin file using your .ino.  Neither approach works, and always fails with the errors similar to above.

Any ideas or suggestions would be greatly appreciated

**Comment by ClaesClaes** [comment]
What you are experience is a weakness in my very rudimentary example code but easily fixed. 

In the ESP .ino file (line 152) the code assumes that the size of the final chunk of data is less than 510 bytes:

`if (rxData.length() < 510) // TODO Asumes at least 511 data bytes (@ble 4.2)`

This tells the ESP that the final chunk of data sent from the iOS device has arrived. It's hard coded to 510 which in the iOS code called chunkSize (line 435 in BluetoothLE.swift). However the chunkSize is variable and depends upon the MTU negotiated between the iOS device and the ESP. So if your 'negotiated' MTU size is less than 513 (and reducing with 3 = 510, to make room for a header added by the BLE stack) you will end up with a chunkSize of less than 510. The ESP code on line 152 then interprets your **first** chunk of data as the **last** chunk, and that causes the error you are seeing.

You see this from line 435 in BluetoothLE.swift:
`chunkSize = discoveredPeripheral.maximumWriteValueLength (for: .withoutResponse) - 3`

So to fix this, put a breakpoint in the iOS code and see what your chunkSize is and adjust accordingly in the ESP .ino file.

As mentioned this code is very basic, so a better way to handle this is to measure the length of the first chunk in the ESP .ino code, store in a variable and then replace the '510' with that variable on line 152.

Hope this helps.

**Comment by aallard** [comment]
Thank you very much for taking the time to reply, and to provide a thorough explanation of the problem.  I had worked through the issue last night and discovered the problem, exactly as you outlined.  When I realized it was flagging the first chunk of data as the last chunk of data, I worked backwards to figure out how that happened and saw how anything less than 510 was assumed to be the final chunk.  I had planned to post another reply to this thread today with my discovery, but you beat me to it :)

Thank you again for being so responsive.  Hopefully this will help others who run across the same issue.
